pulser-core/pulser/sampler/samples.py

```diff
@@ -83,37 +76,91 @@ class ChannelSamples:
     amp: np.ndarray
     det: np.ndarray
     phase: np.ndarray
-
-    slots: list[_TimeSlot] = field(default_factory=list)
+    slots: list[_TargetSlot] = field(default_factory=list)
 
     def __post_init__(self) -> None:
         assert len(self.amp) == len(self.det) == len(self.phase)
+        self.duration = len(self.amp)
 
         for t in self.slots:
             assert t.ti < t.tf  # well ordered slots
-        for t1, t2 in _pairwise(self.slots):
+        for t1, t2 in zip(self.slots, self.slots[1:]):
             assert t1.tf <= t2.ti  # no overlaps on a given channel
 
+    def extend_duration(self, new_duration: int) -> ChannelSamples:
+        """Extends the duration of the samples.
+
+        Args:
+            new_duration: The new duration for the samples (in ns).
+                Must be greater than or equal to the current duration.
+
+        Returns:
+            The extend channel samples.
+        """
+        extension = new_duration - len(self.amp)
+        if new_duration < self.duration:
+            raise ValueError("Can't extend samples to a lower duration.")
+
+        new_amp = np.pad(self.amp, (0, extension))
+        new_detuning = np.pad(self.det, (0, extension))
+        new_phase = np.pad(
+            self.phase,
+            (0, extension),
+            mode="edge" if self.phase.size > 0 else "constant",
+        )
+        return ChannelSamples(new_amp, new_detuning, new_phase, self.slots)
+
+    def modulate(self, channel_obj: Channel) -> ChannelSamples:
+        """Modulates the samples for a given channel.
+
+        Args:
+            channel_obj: The channel object for which to modulate the samples.
+
+        Returns:
+            The modulated channel samples.
+        """
+        if not isinstance(channel_obj, Channel):
+            raise TypeError("'channel_obj' must be a Channel instance.")
```

Comment:
Why having a runtime exception here when we have the type hint to prevent passing something else?
Is this function high enough to be potentially misused by a regular user?

Reply:
Hum, you're right, for now this won't be directly used by a user, I guess I can remove this check.